Add Physics classes for Navier Stokes Weakly Compressible Finite Volume

GiudGiud opened the pull request:
Rebased and deleted the components work from the original physics PR.
So this is essentially piece two out of three of the effort. 

Piece three will probably be more intelligent to bring in AFTER components have been re-written as actions. 
Also if we create a parent module for all fluid work, that work would go there instead

TODO
- [x] Add testing
- [x] Deprecate Modules/NSFV syntax and forward to Physics
- [x] Link physics harder to avoid duplicating parameters.
- [x] Add doco

One issue with linking physics to each other is the order of creation. We cant retrieve them in the constructor unless they are created in order, which we currently only do for mesh generators and it requires quite a bit of code.

A solution would be to retrieve Physics only in the initializePhysicsAdditional(), which is a task happening after construction. Thoughts?

GiudGiud committed:
Add a base class for the WCNSFV flow physics
refs #25642

GiudGiud committed:
Add a base class for all the Navier Stokes flow physics
refs #25642

GiudGiud committed:
Add parameters for the WCNSFV flow equations

GiudGiud committed:
Add parameters for the WCNSFV heat equation

GiudGiud committed:
Add interfaces needed for NSFV at the framework level for restart and FV
- calls to addXXX from the action
- virtual definitions of addXXXX
- forwarding call to addRelationShipManager

Add more parameter checking routines in physics

GiudGiud committed:
Add more APIs at framework Physics level from NSFVBase

GiudGiud committed:
Add variable and boundaries at the NS Physics base class level
- they are physics concepts and are shared between all of our NS implementations

GiudGiud committed:
Easier logic for NSFVBase

GiudGiud committed:
Add kernels and variables in WCNSFV flow physics

GiudGiud committed:
Add BCs in WCNSFV flow physics

GiudGiud committed:
Handle Rhie CHow interpolator in WCNSFV base class

GiudGiud committed:
Add params and kernels for heat advection physics

GiudGiud committed:
Remove templated getParams
add more comments

GiudGiud committed:
Add boundary conditions and ICs for heat advection

GiudGiud committed:
Add boundary conditions for flow physics
requiring postprocessors too

GiudGiud committed:
Add more parameter checking routines, that fit the needs of NS physics

GiudGiud committed:
Make a more robust RC UO unicity check
- the concern is that users may specify more than one flow-physics derived physics
and they all need a RC UO so they all may create it

GiudGiud committed:
Add a test input for the Navier Stokes flow physics

GiudGiud committed:
Add missing parameters at the NS Physics top level
Rework the parameter checking to accomodate for multimooseenum instead of std::vector<enum>

GiudGiud committed:
Add material for heat advection
Use RC uo name routine to keep consistent

GiudGiud committed:
Add missing kernels for NSFlow physics
Use RC UO name routine for consistency
Add missing parameters for numerical discretization of face interpolation of advected quantities
Use new routines for conditional parameter checking

GiudGiud committed:
Add more docs on how to create a Physics object

GiudGiud committed:
Redo ghost layer management, at the input level this time
We are not building an action so we cant get really early access to the number of layers necessary
(unless we hard code them in the parameters, to be conservatie for example)

GiudGiud committed:
Fixup parameters for heat advection

GiudGiud committed:
Add capability to link flow & advection physics

GiudGiud committed:
Add block consistency checks from NSFVAction at physics level

GiudGiud committed:
Adapt RC UO name to make sure we use the flow one

GiudGiud committed:
Check the coupled physics parameters to make sure they match

GiudGiud committed:
Only add time derivative terms for transients
Dont rely on blockrestrictable for kernels

GiudGiud committed:
Remove block restrictable interface from Physics, not capable of tracking blocks with a dynamic setup
Keep track of variable names on a per-physics basis to enable restarting at the base-class level

GiudGiud committed:
Add input file for energy equation

GiudGiud committed:
WIP comment out crashing things

GiudGiud committed:
Better error messages

GiudGiud committed:
Guard against bad mesh

GiudGiud committed:
Add a clearer error message for debugging component ordering

GiudGiud committed:
Move some explanation
Let everyone change the block restriction of physics

GiudGiud committed:
Append Base to NavierStokesFlowPhysics

GiudGiud committed:
Factor out scalar field advection parameters

GiudGiud committed:
Add Navier Stokes scalar advection Physics

GiudGiud committed:
Silence compile warnings

GiudGiud committed:
Add robut parameter checking to WCNSFV physics

GiudGiud committed:
Add more options for checking input parameters
To meet needs of NSFV parameter checking

GiudGiud committed:
Remove default for wall BCs

GiudGiud committed:
Re-organize common parameter checking between Physics
and commonalize the coupled_flow_physics parameter

GiudGiud committed:
Factor out most parameters for momentum scalar and energy equations from NSFVAction
to be able to share them with the respective Physics without transferring them one by one

GiudGiud committed:
Finally figure out how to not overlap RC UOs when defining two physics

GiudGiud committed:
Add an input file for scalar advection in a regular mesh simulation

GiudGiud committed:
Add a verbose boolean to physics to encourage debug print outs by developers

GiudGiud committed:
Add routines to be able to get names of variables from physics easily
Dont flag boundary conditions if there's no issue

GiudGiud committed:
Add input file for solving 2 NS physics that are disconnected

GiudGiud committed:
Rename routine to save variable names

GiudGiud committed:
Factor out turbulence parameters

GiudGiud committed:
Enable compilation with warnings

GiudGiud committed:
Add a turbulence Physics for WCNSFV

GiudGiud committed:
Enable physics action to create aux variable & kernel
Add another parameter error routine

GiudGiud committed:
Add turbulence to scalar physics

GiudGiud committed:
Address compilation warnings

GiudGiud committed:
Remove physics from the problem
Use actions instead

GiudGiud committed:
Debug commit

GiudGiud committed:
Add a way to get a Physics from the action warehouse

GiudGiud committed:
Adapt to Physics being actions

GiudGiud committed:
Adapt to Physics being actions in NS

GiudGiud committed:
Adapt input files to new syntax

GiudGiud committed:
Dont trust the problem for the dimension

GiudGiud committed:
Add more support for multiple turbulence physics

GiudGiud committed:
Add routines to check if parameters are merge-able

GiudGiud committed:
Add a routine to get all the Physics of a certain type from the ActionWarehouse

GiudGiud committed:
Factor out function BCs for inlets as we will use functors instead in Physics

GiudGiud committed:
Add capability to merge parameters from other Physics objects
- move a lot of parameters data to attributes upon construction so we can add to those block- or boundary- restricted attributes

GiudGiud committed:
Adapt test inputs for:
- new functors for inlet/wall parameters
- new syntax for Physics (just the 2 block input)
- new organization of momentum parameters (defined at FlowPhysics only when possible)

GiudGiud committed:
Add more parameter checking on boundaries since they are hard to check
before junctions are created
Consolidate outlet momentum BC code

GiudGiud committed:
Adapting input file for merging Physics and new syntax

GiudGiud committed:
Add multimooseenum header, probably not needed

GiudGiud committed:
Add block-concatenation of friction parameters in wcnsfv flow physics
Add turbulence-physics coupling to flow physics to be able to select the right viscosity

GiudGiud committed:
Adapt more inputs to new functor parameters for inlet/outlet BCs
Adapt inputs to junction syntax that avoids over-specifying things
Add a simple single-physics single-component test

GiudGiud committed:
Add better capabilities for building maps from vectors & multimooseenum

GiudGiud committed:
Make vector inlet and outlet parameters maps so that we dont need to provide values for all boundaries

GiudGiud committed:
Clean up Physics a bit

GiudGiud committed:
Inherit tasks from parent WCNSFV physics class

GiudGiud committed:
Add routine to check block restriction of variables
to physics base class

GiudGiud committed:
Check block restriction of variables in WCNSFV actions

GiudGiud committed:
Use functor BC for heat advection physics for flexibility

GiudGiud committed:
Fixups post-rebase

GiudGiud committed:
Shorten WCNSFV Physics names in input files

GiudGiud committed:
Use Physics on old Modules/NavierStokesFV syntax

GiudGiud committed:
Compatibility adjustements for WCNSFV base physics class for dual syntax

GiudGiud committed:
Compatibility adjustements for WCNSFV flow physics for dual syntax

GiudGiud committed:
Compatibility adjustements for WCNSFV heat advection physics for dual syntax

GiudGiud committed:
Compatibility adjustements for WCNSFV scalar advection physics for dual syntax

GiudGiud committed:
Adapt action error testing to physics use

GiudGiud committed:
Update MMS test syntax as they do not allow deprecation

GiudGiud committed:
Allow -optionally- an empty param when comparing two params

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2136651/) on 5adcd97 wanted to post the following:

The following commit(s) contain fixups:

- ffc98b283e: fixup! Fixups post-rebase

GiudGiud committed:
Compatibility adjustements for WCNSFV turbulence physics for dual syntax

GiudGiud committed:
Request FV problem for WCNSFV Physics

GiudGiud committed:
Additional parameter checking capabilities

GiudGiud committed:
Rework exodus restart physics capability

GiudGiud committed:
Add test inputs for physics + incompressible flow

GiudGiud committed:
Use a separate task for copying variables from Exodus
Otherwise all the physics get auto-built? this is not right

GiudGiud committed:
Better heuristic for detecting use of the old action syntax

GiudGiud commented:
This is probably ready enough to be worth looking at @grmnptr @lindsayad 
I can probably trim down the duplication in parameters a bit, this might the sorest issue right now but I'd be happy to address others you may find

GiudGiud committed:
Rename base class for NS physics

GiudGiud committed:
Add example inputs for the overview of the system

GiudGiud committed:
Add documentation to the Physics

GiudGiud committed:
Always have a flow physics defined, rely on it for synchronizing many parameters
- flux bcs values and directions
- variable face interpolation
- misc parameters

GiudGiud committed:
Add ability to fish for a Physics of the right type
And to check it block restriction

GiudGiud committed:
Remove NS base class and move attributes to flow class
retrieve those attributes (boundaries etc) from the flow physics

GiudGiud committed:
Remove WCNS base class, re-introduce a NS base class that is very general
Instead of calling _flow_physics->getTHIS everywhere, use a helper class to hold attributes

GiudGiud committed:
Adapt test syntax

grmnptr commented on framework/include/physics/PhysicsBase.h:
```diff
@@ -249,3 +463,177 @@ PhysicsBase::checkVectorParamsNoOverlap(const std::vector<std::string> & param_v
+bool
+PhysicsBase::parameterConsistent(const InputParameters & other_param,
+                                 const std::string & param_name,
+                                 bool warn) const
```
```suggestion
                                const bool warn) const
```

grmnptr commented on framework/include/physics/PhysicsBase.h:
```diff
@@ -249,3 +463,177 @@ PhysicsBase::checkVectorParamsNoOverlap(const std::vector<std::string> & param_v
+PhysicsBase::warnInconsistent(const InputParameters & other_param,
+                              const std::string & param_name) const
+{
+  parameterConsistent<T>(other_param, param_name, true);
```
I think you should lift the warning from parameter consistent to this guy and let parameter consistent just decide if it is consistent or not.

GiudGiud replied:
removed the consistency checks. I added that to merge physics over multiple components. I dont know if this will be used in the end

grmnptr commented on framework/include/physics/PhysicsBase.h:
```diff
@@ -249,3 +463,177 @@ PhysicsBase::checkVectorParamsNoOverlap(const std::vector<std::string> & param_v
+  {
+    return map;
+    // If we cant return a map of default C, dont try it
+    // if constexpr (std::is_same_v<MooseEnum, T> || std::is_same_v<MultiMooseEnum, T>)
+    //   return map;
+
+    // C def;
+    // for (const auto & k : keys)
+    //   map[k] = def;
+    // return map;
+  }
+  std::transform(keys.begin(),
+                 keys.end(),
+                 values.begin(),
+                 std::inserter(map, map.end()),
+                 [](T a, C b) { return std::make_pair(a, b); });
+  return map;
+}
+
+template <typename T>
+std::map<T, MooseEnum>
+PhysicsBase::createMapFromVectorAndMultiMooseEnum(std::vector<T> keys, MultiMooseEnum values) const
+{
+  std::map<T, MooseEnum> map;
+  // No values have been specified. We cant form a map of empty MooseEnum
+  if (!values.size())
+    return map;
+  std::transform(keys.begin(),
+                 keys.end(),
+                 values.begin(),
+                 std::inserter(map, map.end()),
+                 [values](T a, MooseEnumItem b)
+                 {
+                   // Create a MooseEnum from the available values in the MultiMooseEnum and an
+                   // actual current active item from that same MultiMooseEnum
+                   MooseEnum single_value(values.getRawNames(), b.name());
+                   return std::make_pair(a, single_value);
+                 });
+  return map;
+}
```
I think these should go to moose one of the utility files.

lindsayad replied:
We should be able to pass in the function arguments as const references and not do all these container copies

GiudGiud replied:
fixed

grmnptr commented on framework/include/utils/MooseMeshUtils.h:
```diff
@@ -289,4 +289,19 @@ void makeOrderedNodeList(std::vector<std::pair<dof_id_type, dof_id_type>> & node
+void stitchNodesets(MooseMesh & input_mesh,
+                    const BoundaryName & side1,
+                    const BoundaryName & side2,
+                    bool allow_partial,
+                    bool keep_old);
```
```suggestion
                    const bool allow_partial,
                    const bool keep_old);
```

grmnptr commented on framework/src/base/MooseApp.C:
```diff
@@ -1525,16 +1525,16 @@ MooseApp::run()
-  {
-    mooseError(err.what());
-  }
+  // try
+  // {
+  // TIME_SECTION("setup", 2, "Setting Up");
+  setupOptions();
+  runInputFile();
+  // }
+  // catch (std::exception & err)
+  // {
+  //   mooseError(err.what());
+  // }
```
Man, this is a really annoying try that you have to disable for debugging. Once you are in a good shape and you don't need the debugger anymore, make sure you reset this guy.

grmnptr commented on framework/src/physics/PhysicsBase.C:
```diff
@@ -127,20 +129,28 @@ PhysicsBase::act()
     addExecutors();
 
+  // Exodus restart capabilities
+  if (_current_task == "copy_nodal_vars_physics")
```
```suggestion
  if (_current_task == "copy_vars_physics")
```

grmnptr replied:
Considering that the function changed too.

grmnptr commented on framework/include/physics/PhysicsBase.h:
```diff
@@ -96,14 +122,59 @@ class PhysicsBase : public Action
   /// Check that a parameter is set only if the first one is set to true
   void checkSecondParamSetOnlyIfFirstOneTrue(const std::string & param1,
                                              const std::string & param2) const;
```
I would move all of these into an `inputparamsutils.h/c`, feels like they are very general and could be reused.

GiudGiud replied:
There s diamond inheritance problems to deal with because of the need for paramError but I ll see what I can do

lindsayad replied:
could consider virtual inheritance

GiudGiud replied:
I think we should yes. I ll look at it more seriously this could simplify things

but this parameter checking utility that is going away when we have functor-parameter-checking is not the place

grmnptr reviewed (changes requested):
Posting through the framework.

GiudGiud commented on framework/include/physics/PhysicsBase.h:
```diff
@@ -185,19 +269,33 @@ class PhysicsBase : public Action
 template <typename T>
 const T *
-PhysicsBase::getCoupledPhysics(const PhysicsName & phys_name) const
+PhysicsBase::getCoupledPhysics(const PhysicsName & phys_name, const bool allow_fail) const
```
we are looking for a physics but it does not exist. for example a flow case without turbulence

GiudGiud commented on framework/include/physics/PhysicsBase.h:
```diff
@@ -233,6 +331,122 @@ PhysicsBase::checkVectorParamsSameLength(const std::string & param1,
+void
+PhysicsBase::checkTwoDVectorParamMultiMooseEnumSameLength(const std::string & param1,
+                                                          const std::string & param2,
+                                                          const bool error_for_param2) const
```
sometimes, saying :
param1 is not the same size as param2 is clearer

because for example the number of inlet boundaries not being the same as the number of inlet functors points to the inlet when the functors are probably the problem

GiudGiud committed:
Address Peter's review
- add docstrings
- separate utilities for checking parameters
- rename variable loading task

Co-authored-by: Peter German <31662443+grmnptr@users.noreply.github.com>

GiudGiud committed:
Dont forget to retrieve the turbulence physics from the flow one

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2147064/) on 46bb6a1 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/27169/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/27169/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 22568448d90bc6db05f68015c0a1ecdf24245bfe`

GiudGiud committed:
Separate parameter checking utility into a separate utils

GiudGiud commented on framework/include/physics/PhysicsBase.h:
```diff
@@ -126,6 +197,18 @@ class PhysicsBase : public Action
+  /// Check if a vector contains all the mesh blocks
+  bool allMeshBlocks(const std::vector<SubdomainName> & blocks) const;
+
+  /// Routine to help create maps
+  template <typename T, typename C>
+  std::map<T, C> createMapFromVectors(std::vector<T> keys, std::vector<C> values) const;
+  template <typename T>
+  std::map<T, MooseEnum> createMapFromVectorAndMultiMooseEnum(std::vector<T> keys,
+                                                              MultiMooseEnum values) const;
```
I ll have to think of a name. Maybe StdMapUtils ?

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2147493/) on 37fc6dc wanted to post the following:

The following commit(s) contain fixups:

- f0c6f71843: fixup! Debug commit

GiudGiud committed:
Add docstrings

GiudGiud commented:
ok most of the first set of comments should be addressed. I may or may not be able to address more from Ecuador but dont let that stop you

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2152216/) on 4b42a9a wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/27169/site)

This comment will be updated on new commits.

GiudGiud committed:
Clean up documentation and add index pages (links) for syntax

GiudGiud committed:
Address test failures:
- component work is not ready
- NSFVaction always creates turbulence physics, but turbulence isnt always active

GiudGiud committed:
Add doco for deprecated action

GiudGiud committed:
Cleaer log for functor name fail in WCNSFV heat advection
More addressing test failures
- remove duplicate test requirement
- fix snippet links
- use new THM enum in tests output
- propagate block restrictions silently from flow physics to other physics

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2156836/) on 65904b1 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/27169/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/27169/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format b02bb194f5ca6361a22e45948575b03c3d10f698`

lindsayad commented on modules/navier_stokes/test/tests/finite_volume/ins/channel-flow/2d-mixing-length-physics.i:
```diff
@@ -0,0 +1,89 @@
+    []
+    [WCNSFVScalarAdvection]
+      [scalars]
+        add_scalar_equation = true
```
It seems a little redundant to have `add_scalar_equation` two lines after the `WCNSFVScalarAdvection` label. Is there anything we can do about that?

lindsayad replied:
looking at your other test inputs, it seems like this parameter is not necessary which is great

GiudGiud replied:
I ll do a sweep to remove those unnecessary ones. I started out with something more verbose than now

GiudGiud replied:
done

lindsayad commented on modules/navier_stokes/test/tests/finite_volume/ins/channel-flow/2d-rc-transient-physics.i:
```diff
@@ -0,0 +1,97 @@
+        density = 'rho'
+        dynamic_viscosity = 'mu'
+
+        initial_velocity = '${u_inlet} 1e-12 0'
```
Can we just do 0 instead of 1e-12? In the old days we had to do stuff like this so we didn't divide by a zero velocity norm, but we should be past that

GiudGiud replied:
done

lindsayad commented:
so are we keeping the non-physics action?

GiudGiud commented:
> so are we keeping the non-physics action?

I can remove it. A good chunk of its code is used in the filemeshwcnsfvcomponent in thm. So I cant remove it until we remove that component. I can remove its user facing side

lindsayad commented:
it would be nice to reduce the lines-of-code maintenance burden at some point for sure

GiudGiud committed:
Dont seek description for NSFVAction that is no longer registered to Modules/NavierStokesFV syntax

GiudGiud committed:
More includes for unity build

grmnptr commented:
To be honest, I'd say we should ask the users first about the removal. But it looks like the syntax is almost the same so I'd be in favor of removing the old one. At the same time, I haven't reviewed the module version yet, I'll try to do it ASAP.

GiudGiud commented:
We can support the old syntax. And we can make the new syntax more like the old syntax (all in 1 block) if you all prefer that.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2167639/) on c2d4d81 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>c7c219</th>
      <th colspan=4>#27169 c2d4d8</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.14%</td>
      <td>85.08%</td>
      <td>-0.06%</td>
      <td>32.17%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>102691</td>
      <td>102715</td>
      <td>+24</td>
      <td>46</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>17930</td>
      <td>18016</td>
      <td>+86</td>
      <td>97</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/27169/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/27169/coverage/framework)


## Modules coverage

### Navier stokes

<table>
  <thead>
    <tr>
      <th></th>
      <th>c7c219</th>
      <th colspan=4>#27169 c2d4d8</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>84.79%</td>
      <td>77.75%</td>
      <td>-7.04%</td>
      <td>93.49%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>14613</td>
      <td>14665</td>
      <td>+52</td>
      <td>1593</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>2621</td>
      <td>4197</td>
      <td>+1576</td>
      <td>111</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/27169/coverage_diff/navier_stokes)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/27169/coverage/navier_stokes)

### Thermal hydraulics

<table>
  <thead>
    <tr>
      <th></th>
      <th>c7c219</th>
      <th colspan=4>#27169 c2d4d8</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>89.11%</td>
      <td>89.08%</td>
      <td>-0.02%</td>
      <td>58.33%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>13969</td>
      <td>13974</td>
      <td>+5</td>
      <td>7</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>1708</td>
      <td>1713</td>
      <td>+5</td>
      <td>5</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/27169/coverage_diff/thermal_hydraulics)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/27169/coverage/thermal_hydraulics)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/27169/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/27169/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/27169/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/27169/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/27169/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/27169/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/27169/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/27169/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/27169/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/27169/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/27169/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/27169/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/27169/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/27169/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/27169/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/27169/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/27169/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/27169/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/27169/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/27169/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/27169/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/27169/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/27169/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/27169/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/27169/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/27169/coverage/stochastic_tools)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/27169/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/27169/coverage/xfem)

</details>

## Warnings

- `framework` new line coverage rate 32.17% is less than the suggested 90.0%
- `thermal_hydraulics` new line coverage rate 58.33% is less than the suggested 90.0%



This comment will be updated on new commits.

GiudGiud committed:
Docstring improvements and remove unused routines

joshuahansel commented:
I bet @GiudGiud really wants to cash in that free approve coupon here

GiudGiud commented on framework/include/utils/InputParametersChecksUtils.h:
```diff
@@ -0,0 +1,549 @@
+  // MooseObjects and Actions (MooseParameterInterface-derived classes)
+  /// Forwards parameter check to the class using this utility
+  template <typename T>
+  T forwardGetParam(const std::string & param_name) const
```
do you like these or do you think it s obnoxious and I should just have the `_customer_class->template getParam<T>(param_name);` directly in there

joshuahansel replied:
I know it's ugly when you have to do `this->template blah<T>()`, but what about ones like `forwardIsParamValid`: what's the reason for those?

GiudGiud replied:
uniformity?

joshuahansel replied:
Could these maybe go into `MooseBaseParameterInterface`?

GiudGiud replied:
these few routines forward to routines in `MooseBaseParameterInterface`. we dont particularly need them outside of this parameter checking class, which is why they are `private`

grmnptr commented on modules/navier_stokes/src/physics/WCNSFVScalarAdvectionPhysics.C:
```diff
@@ -0,0 +1,385 @@
+            ") is not equal to the number of inlet boundaries (" +
+            std::to_string(inlet_boundaries.size()) + ") times the number of passive scalars (" +
+            std::to_string(_passive_scalar_names.size()) + ")");
+  if (_passive_scalar_names.size() != _passive_scalar_inlet_functors.size())
+    paramError("passive_scalar_inlet_functors",
+               "The number of groups of inlet functors (" +
+                   std::to_string(_passive_scalar_inlet_functors.size()) +
+                   ") is not equal to the number of passive scalars (" +
+                   std::to_string(_passive_scalar_names.size()) + ")");
```
Maybe you can replace this with you new parameter checks.

GiudGiud replied:
components havent had the chance to modify the boundaries at this point
so we must wait till last minute, and cant rely on parameters only

grmnptr commented on modules/navier_stokes/src/physics/WCNSFVScalarAdvectionPhysics.C:
```diff
@@ -0,0 +1,385 @@
+
+  for (const auto name_i : index_range(_passive_scalar_names))
+  {
+    if (inlet_boundaries.size() != _passive_scalar_inlet_functors[name_i].size())
+      paramError("passive_scalar_inlet_functors",
+                 "The number of inlet boundary functors for scalar '" +
+                     _passive_scalar_names[name_i] +
+                     "' does not match the number of inlet boundaries (" +
+                     std::to_string(_passive_scalar_inlet_functors[name_i].size()) + ")");
```
MAybe you can replace this too with the new checks.

GiudGiud replied:
same as other comment, resolving to minimize items on PR

grmnptr commented on modules/navier_stokes/src/physics/WCNSFVTurbulencePhysics.C:
```diff
@@ -0,0 +1,297 @@
+registerMooseAction("NavierStokesApp", WCNSFVTurbulencePhysics, "add_material");
+
+InputParameters
+WCNSFVTurbulencePhysics::validParams()
```
Nice, I like this a lot

grmnptr commented on modules/navier_stokes/test/tests/finite_volume/ins/channel-flow/2d-mixing-length-physics.i:
```diff
@@ -0,0 +1,89 @@
+  []
+[]
+
+[Physics]
+  [NavierStokes]
+    [WCNSFVFlow]
+      [flow]
+        compressibility = 'incompressible'
+
+        density = ${rho}
+        dynamic_viscosity = ${mu}
+
+        initial_velocity = '1 1 0'
+        initial_pressure = 0.0
+
+        inlet_boundaries = 'left'
+        momentum_inlet_types = 'fixed-velocity'
+        momentum_inlet_function = '1 0'
+
+        wall_boundaries = 'top bottom'
+        momentum_wall_types = 'noslip symmetry'
+
+        outlet_boundaries = 'right'
+        momentum_outlet_types = 'fixed-pressure'
+        pressure_function = '0'
+
+        momentum_advection_interpolation = ${advected_interp_method}
+        mass_advection_interpolation = ${advected_interp_method}
+      []
+    []
```
Hey @tanoret , what do you think about this syntax? Can you envision your current two-phase/turbulence routines in it? I think it is pretty flexible.

GiudGiud replied:
see the whole input, this only shows flow

tanoret replied:
Normally for the two-phase flow we will need to add mixture materials as materials, which is doable by extending the current actions, but it would be much harder to set up the kernels fo each case required for two-phase flow. Also setting up solidification and melting will be quite hard in this configuration...

GiudGiud replied:
If the equations for flow are close enough, then I can see if we can make do with using the same Physics object

likely not though. I think we will want to have a dedicated two phase block "[WCNSFV2PhaseFlow]" for example

grmnptr reviewed (changes requested):
Nice! Just a few comments/questions.

GiudGiud commented on modules/navier_stokes/include/base/NSFVBase.h:
```diff
@@ -767,15 +699,197 @@ NSFVBase<BaseType>::validParams()
+
+  params += NSFVBase<BaseType>::commonMomentumEquationParams();
+
+  params.addParam<MooseFunctorName>(
+      "dynamic_viscosity", NS::mu, "The name of the dynamic viscosity");
+
+  params.addParam<MooseFunctorName>("density", NS::density, "The name of the density");
```
moving it down

GiudGiud commented on modules/navier_stokes/include/physics/WCNSFVScalarAdvectionPhysics.h:
```diff
@@ -0,0 +1,77 @@
+
+  unsigned short getNumberAlgebraicGhostingLayersNeeded() const override;
+
+  // /**
+  //  * Functions adding kernels for the incompressible / weakly compressible scalar advection
+  //  * equation
+  //  * If the material properties are not constant, some of these can be used for
+  //  * weakly-compressible simulations as well.
+  //  */
```
```suggestion
   /**
    * Functions adding kernels for the incompressible / weakly compressible scalar advection
    * equation
    * If the material properties are not constant, some of these can be used for
    * weakly-compressible simulations as well.
    */
```

GiudGiud commented on modules/navier_stokes/src/physics/WCNSFVFlowPhysics.C:
```diff
@@ -0,0 +1,1406 @@
+  if (isParamValid("coupled_turbulence_physics"))
+    return getCoupledPhysics<WCNSFVTurbulencePhysics>(
+        getParam<PhysicsName>("coupled_flow_physics"));
+  // Look for any physics of the right type, and check the block restriction
+  else
+  {
```
it got added in the input as a Physics. they just didnt specify it as the parameter 'coupled_turbulence_physics'

the results would be different. you d get most of the kernels, but not the use of the right total viscosity in the BCs of the flow physics

GiudGiud committed:
Address Peter's review

- rename physics classes

Co-authored-by: Peter German <31662443+grmnptr@users.noreply.github.com>

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2185028/) on d71a9dc wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/27169/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/27169/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format a489a53bdb58ce7053c80cbebf6e2673da28fa3b`

grmnptr commented on modules/navier_stokes/include/base/NSFVBase.h:
```diff
@@ -767,15 +699,197 @@ NSFVBase<BaseType>::validParams()
+   */
+  params += NSFVBase<BaseType>::commonScalarFieldAdvectionParams();
+
+  // These parameters are not shared because the WCNSFVPhysics use functors
```
I opened this:
https://github.com/idaholab/moose/issues/27408

Turns out @tanoret needs this for a training as soon as possible so we'll add this very fast so we can get a BlueCrab version for the trainees asap.

GiudGiud replied:
rebased and it s inclluded now

GiudGiud committed:
Rename syntax in tests

lindsayad commented:
My implicit approval is given when Peter approves 😄

GiudGiud commented on modules/navier_stokes/doc/content/source/physics/WCNSFVFlowPhysics.md:
```diff
@@ -0,0 +1,60 @@
+
+## Coupling with other Physics
+
+The heat advection equation can be solved concurrently with the flow equations using an additional [WCNSFVHeatAdvectionPhysics.md].
```
changed

GiudGiud commented on framework/src/utils/InputParametersChecksUtils.C:
```diff
@@ -0,0 +1,10 @@
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "InputParametersChecksUtils.h"
```
removed

GiudGiud commented on modules/navier_stokes/include/physics/WCNSFVFlowPhysics.h:
```diff
@@ -0,0 +1,187 @@
+  /// Get the inlet flux postprocessor if using a flux inlet
+  const std::vector<PostprocessorName> & getFluxInletPPs() const { return _flux_inlet_pps; }
+  /// Return the name of the Rhie Chow user object
+  std::string rhieChowUOName() const;
```
done

GiudGiud commented on modules/navier_stokes/include/physics/WCNSFVScalarAdvectionPhysics.h:
```diff
@@ -0,0 +1,77 @@
+  //  * equation
+  //  * If the material properties are not constant, some of these can be used for
+  //  * weakly-compressible simulations as well.
+  //  */
```
not really

GiudGiud commented on modules/navier_stokes/src/physics/WCNSFVFlowPhysics.C:
```diff
@@ -0,0 +1,1406 @@
+      "Define the Navier Stokes weakly-compressible mass and momentum equations");
+
+  params += NSFVAction::commonMomentumEquationParams();
+  params.addParam<bool>("add_flow_equations",
+                        true,
+                        "Whether to add the flow equations. This parameter is not necessary when "
+                        "using the Physics syntax");
```
scalar-advection only solves. we used that in multiapps with the fully coupled approach to make it more tractable

GiudGiud commented on modules/navier_stokes/src/physics/WCNSFVFlowPhysics.C:
```diff
@@ -0,0 +1,1406 @@
+
+  // Spatial discretization scheme
+  // Specify the numerical schemes for interpolations of velocity and pressure
+  params.transferParam<MooseEnum>(NSFVAction::validParams(), "velocity_interpolation");
```
however it is in NSFVAction currently

GiudGiud commented on modules/navier_stokes/src/physics/WCNSFVFlowPhysics.C:
```diff
@@ -0,0 +1,1406 @@
+}
+
+unsigned short
+WCNSFVFlowPhysics::getNumberAlgebraicGhostingLayersNeeded() const
```
so far yes though we need a layer of geometric too for FV

GiudGiud committed:
Fixups in fluid heat transfer physics

GiudGiud committed:
Better deprecation message for NavierStokesFV

tanoret commented on modules/navier_stokes/src/physics/WCNSFVTurbulencePhysics.C:
```diff
@@ -0,0 +1,328 @@
+  return params;
+}
+
+WCNSFVTurbulencePhysics::WCNSFVTurbulencePhysics(const InputParameters & parameters)
+  : NavierStokesPhysicsBase(parameters),
+    WCNSFVCoupledAdvectionPhysicsHelper(parameters, this),
+    _turbulence_model(getParam<MooseEnum>("turbulence_handling")),
+    _mixing_length_name(getParam<AuxVariableName>("mixing_length_name")),
+    _mixing_length_walls(getParam<std::vector<BoundaryName>>("mixing_length_walls"))
```
This is hardcoded to mixing length. What about k-epsilon? Normally we should have a turbulent model selector as we do in most CFD codes

GiudGiud replied:
We can rename when it gets into the syntax.

GiudGiud replied:
renamed the attribute

tanoret commented on modules/navier_stokes/src/physics/WCNSFVTurbulencePhysics.C:
```diff
@@ -0,0 +1,328 @@
+    params.set<MooseFunctorName>(NS::density) = _density_name;
+    params.set<MooseFunctorName>(NS::cp) = _fluid_energy_physics->getSpecificHeatName();
+    params.set<MooseFunctorName>(NS::mixing_length) = _mixing_length_name;
+    params.set<Real>("schmidt_number") = getParam<Real>("turbulent_prandtl");
```
Why the Schmidt number gets the turbulent Pr? This should be the turbulent Sc number, which is generally quite different than the turbuelnt Pr

GiudGiud replied:
I'll have to check.

GiudGiud replied:
this kernel is being used on energy. this is correct

tanoret commented on modules/navier_stokes/src/physics/WCNSFVTurbulencePhysics.C:
```diff
@@ -0,0 +1,328 @@
+    const auto & passive_scalar_names = _scalar_transport_physics->getAdvectedScalarNames();
+    const auto & passive_scalar_schmidt_number =
+        getParam<std::vector<Real>>("passive_scalar_schmidt_number");
+    if (passive_scalar_schmidt_number.size() != passive_scalar_names.size())
+      paramError("passive_scalar_schmidt_number",
+                 "The number of Schmidt numbers defined is not equal to the number of passive "
+                 "scalar fields!");
+
```
User should be able to define one Sc number for all species, e.g., delay neutron precursor families are genenrally assigned the same Sc number

GiudGiud replied:
adding this as a new feature

tanoret commented on modules/navier_stokes/src/physics/WCNSFVTurbulencePhysics.C:
```diff
@@ -0,0 +1,328 @@
+  params.addClassDescription("Define a turbulence model for a weakly-compressible Navier Stokes "
+                             "flow with a finite volume discretization");
+
+  MooseEnum turbulence_type("mixing-length none", "none");
```
Should also include k epsilon. Models are alreay merged and tested

GiudGiud replied:
Will have to be in a follow up. 
It s bad practice to lump tons of work together, it makes reviewing harder

tanoret commented on modules/navier_stokes/src/physics/WCNSFVTurbulencePhysics.C:
```diff
@@ -0,0 +1,328 @@
+{
+  InputParameters params = NavierStokesPhysicsBase::validParams();
+  params += WCNSFVCoupledAdvectionPhysicsHelper::validParams();
+  params.addClassDescription("Define a turbulence model for a weakly-compressible Navier Stokes "
+                             "flow with a finite volume discretization");
```
Why weakly-compressible only, this one should support incompressible too

GiudGiud replied:
I ll change

tanoret commented on framework/include/actions/ActionWarehouse.h:
```diff
@@ -134,6 +135,18 @@ class ActionWarehouse : public ConsoleStreamInterface
     }
     return *p;
   }
+  template <class T>
```
Why this function needs to be templated, wouldn't it always be returning a `PhysicsBase` or derived class?

GiudGiud replied:
Thanks to the template you can get the specific type you need

This saves you a dynamic cast which is slow

tanoret commented on framework/include/actions/ActionWarehouse.h:
```diff
@@ -157,6 +170,28 @@ class ActionWarehouse : public ConsoleStreamInterface
+  /**
+   * Retrieve all Physics with a specific type ordered by their names
+   */
+  template <class T>
```
Why we need to use this function with templates? I could be using this function with a `getPhysics<Real>`, which makes no sense. If template must be preserved, please check that the actual class on which the function is templated makes sense

GiudGiud replied:
If you use it with the wrong type, it won't find any physics to return

GiudGiud replied:
I ll add an assert though it should help

grmnptr commented on modules/navier_stokes/include/physics/WCNSFVHeatAdvectionPhysics.h:
```diff
@@ -0,0 +1,92 @@
+  /// Energy wall boundary types
+  MultiMooseEnum _energy_wall_types;
+  /// Functors describing the wall boundary values. See energy_wall_types for what the functors actually represent
+  std::vector<MooseFunctorName> _energy_wall_functors;
```
Ahhh, okay thanks for the explanation!

grmnptr commented on modules/navier_stokes/include/physics/WCNSFVScalarAdvectionPhysics.h:
```diff
@@ -0,0 +1,77 @@
+  /// Passive scalar inlet boundary types
+  MultiMooseEnum _passive_scalar_inlet_types;
+  /// Functors describing the inlet boundary values. See passive_scalar_inlet_types for what the functors actually represent
+  std::vector<std::vector<MooseFunctorName>> _passive_scalar_inlet_functors;
```
Ahhh nice, thanks!

GiudGiud committed:
Keep old syntax in use until May 6 training
then we will transition to Physics

GiudGiud committed:
Adapt to boundary parameters being functors now

GiudGiud committed:
Address Mauricio's review:
- add support for specifying only one schmidt number if desired
- check that the type used for templating works
- use turbulence instead of mxiing length for turbulence walls name

GiudGiud commented:
Rebased and reviews should be addressed. I ll work on getting green again now that we are using the old action for most inputs

GiudGiud committed:
Use shorter syntax for physics

grmnptr reviewed (dismissed):
Even fewer questions/comments.

GiudGiud commented on modules/navier_stokes/src/actions/NSFVActionDeprecation.C:
```diff
@@ -0,0 +1,34 @@
+//* This file is part of the MOOSE framework
```
it s not registered. it s not doing anything

GiudGiud committed:
Address Peter's review
- scalar adv -> transport
- fix index of physics obj from NS
- move map utilities to a utility
- change a link from an input that was removed
Apply suggestions from code review

Co-authored-by: Peter German <31662443+grmnptr@users.noreply.github.com>

GiudGiud commented on modules/navier_stokes/doc/content/source/physics/WCNSFVFlowPhysics.md:
```diff
@@ -0,0 +1,60 @@
+
+!alert! note
+
+This physics only supports Rhie-Chow interpolation for the determination
+of face velocities in the advection terms. The face interpolation of the
```
yeah it s probably to match some gold

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2202348/) on e02a5f1 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/27169/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/27169/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format b74351f2d1885d0f5364e913a5d4a60153b3f039`

GiudGiud commented on framework/include/physics/PhysicsBase.h:
```diff
@@ -86,33 +113,32 @@ class PhysicsBase : public Action
+  template <typename T, typename C>
+  std::map<T, C> createMapFromVectors(std::vector<T> keys, std::vector<C> values) const;
+  template <typename T>
+  std::map<T, MooseEnum> createMapFromVectorAndMultiMooseEnum(const std::vector<T> & keys,
```
done

grmnptr commented on modules/navier_stokes/doc/content/source/physics/WCNSFVFlowPhysics.md:
```diff
@@ -0,0 +1,60 @@
+# WCNSFVFlowPhysics
```
I am just looking at it from a user's point of view. I am on the source site and see the available options. I click on WCNSFVFlow and think oh okay so this is what I have to add. But no, it is actually just Flow. I'd advocate for consistency one way or another. What was the conclusion here, @lindsayad ?
Also, I think there might be an alias option in registration, not sure that is possible/useful in this case.

lindsayad replied:
I don't see someone writing a physics for CNSFV or finite elements any time soon from a practical standpoint. I agree that this should be consistent. And I think it should be `Flow`. This is where we're spending the overwhelming majority of our resources and where we'll be spending it in the foreseeable future. We should have a canonical flow object. I think it's ok if in in the future someone comes along and writes `CompressibleFlow` that lives alongside `Flow`. Or `HDGFlow` that lives alongside `Flow`

GiudGiud replied:
The page has been renamed to "Navier Stokes Flow / WCNSFVFlowPhysics"
The syntax in the input has been changed from [WCNSFVFlow] to [Flow]

Given the large differences between fully coupled and segregated setups, we will likely have to write another physics (for readability) in a shorter time frame than expected. This probably means a NSFVFlowPhysicsBase, with WCNSFV and "some segregated prefix" FlowPhysics derived classes. Renaming WCNSFVFlowPhysics in the code now is not time well spent

grmnptr replied:
Okay, if the documentation and the syntax are consistent it is good enough for me. Thanks for the changes! Yes, we should add the segregated solver at some point. I thought we could have added it to the same physics with incorporating the executioner but if you want to avoid large physics we can probably write a new one.

GiudGiud replied:
If you keep the kernels with rougjly the same parameters in the segregated solver we can shimmy something that will be reasonable to maintain.

GiudGiud replied:
It would be my preference. We ll have to look at how different the input file looks like really. The action (physics) really mirror the input file

GiudGiud replied:
I guess I'm pretty inconsistent on this. We'll see!

lindsayad replied:
A few `git mv` and then one `git sed WCNSFVFlowPhysics Flow`. Should take less than 5 minutes. And then the next physics could be `FlowSegregated` or whatever. Or if we think the `FlowSegregated` will the most used, then the current one could be `FlowMonolithic` and the next could be `Flow`. But I won't make this a hill that I'll die on. One thing that does need to be fixed though is that `/` characters do not get rendered on the source index page. I would suggest removing "/ WCNSFVFlowPhysics" from "Navier Stokes Flow / WCNSFVFlowPhysics"

GiudGiud replied:
if it shows as "Navier Stokes Flow WCNSFVFlowPhysics" it's still fine for me

do you have an issue for / not rendering? I might create one I think we should address it this summer

lindsayad replied:
I did not see one. I created https://github.com/idaholab/moose/issues/27525. Ok stubborn goose, we'll keep things as they are

GiudGiud commented:
Review should be addressed.

grmnptr commented:
My comments have been addressed, ready to approve. How about @tanoret , @joshuahansel ?

joshuahansel commented on modules/navier_stokes/test/tests/finite_volume/wcns/channel-flow/2d-transient-physics-energy.i:
```diff
@@ -0,0 +1,139 @@
+[]
+
+[Physics]
+  [NavierStokes]
```
Personally I'm not a fan of the nesting under `NavierStokes` - seems redundant with `WCNSFVFlow`, but it's ok if you feel strongly about it.

joshuahansel replied:
I'm confused - this input file has `Physics/NavierStokes/WCNSFVFlow/*`, but I don't see that syntax registered, only `Physics/NavierStokes/Flow/*`.

joshuahansel replied:
Input file aintgotno test!?

GiudGiud replied:
probably so

GiudGiud replied:
it's a duplicate of the other one checked in. I ll remove it now

GiudGiud replied:
deleted from history

joshuahansel replied:
But I realize this is the trend you've set with your previous physics - I'll live.

joshuahansel commented:
Does github reset my x/66 files viewed back to 0/66 every time you force-push?

GiudGiud commented:
> Does github reset my x/66 files viewed back to 0/66 every time you force-push?

it s not supposed to. However, I experimented with git filter-branch so every single file did get reset for a moment.
but then I changed my mind and did a classic rebase to delete that file from history

joshuahansel commented:
It's happened to me twice when reviewing your PR :(

GiudGiud commented:
every time I edit history before what you have reviewed I think that would cause the reset.
the solution would be for me to only add new commits ?

joshuahansel commented:
I think best practice is to avoid changing history whenever your PR is no longer draft, maybe until final cleanup after review?

GiudGiud commented on framework/include/utils/MapConversionUtils.h:
```diff
@@ -0,0 +1,73 @@
+                 keys.end(),
+                 values.begin(),
+                 std::inserter(map, map.end()),
+                 [](const T & a, const C & b) { return std::make_pair(a, b); });
```
not everything makes it. I ll assert that, good diea

GiudGiud replied:
fixed

GiudGiud commented on modules/navier_stokes/test/tests/finite_volume/wcns/channel-flow/2d-transient-physics.i:
```diff
@@ -0,0 +1,135 @@
+  [NavierStokes]
+    [Flow]
+      [flow]
+        compressibility = 'weakly-compressible'
```
incompressible is set with the "compressibility" parameter

for others, for example compressible, then we'll have to figure out something

lindsayad commented:
In a meeting with some analysts on Monday, they wanted something simple like `Flow`. To them `WCNSFV` is meaningless. I too like the matching of the syntax with the object, but I am also a dirty dev. I think it's fair if they want something simple like `Flow`. Like if you want to build a model in commercial software X, you don't have to parse an acronym or some really long name

joshuahansel commented:
Fair enough, I figured that was the case; `Flow` is the right choice.

GiudGiud commented:
>  it should be NavierStokesFlowPhysics or NSFlowPhysics and should encompass all of the flow alternatives

the problem is that is going to be huge unmaintainable objects, for little to no gain since the WCNSFV (coupled and segregated, using same kernels so far but soon wont be) are the only one producing results right now

FlowCG / FlowHDG / FlowSegregated (defaulting to segregated) seems a fine expansion strategy to me.
I would prefer a "discretization = " parameter but we would need to change the way actions work for that.

lindsayad commented:
> the only one producing results right now

I don't know about that. Finite volume still struggles mightily to do ALE whereas finite element does not. Additionally, I have not seen our finite volume perform very well on buoyancy driven flows whereas FE doesn't have a problem. It may be the only one producing results in the NEAMS TH area because it's the only discretization used in the NEAMS TH area. But I don't care to dwell on this. The TH area has made it's choice and I support it. In that vein, I think it's fine for `Flow` to represent only incompressible/weakly-compressible finite volume discretizations. It's the industry standard

GiudGiud commented:
I think Sinan's cask results were buoyancy results, in finite volume and were much better than what we did 3 years prior with FE. Though there was a lot of effort to get those cask results.

joshuahansel commented:
Sounds good, we should be able to expand without disrupting users if WC is the "default".
